### PR TITLE
Vector to random points

### PR DESCRIPTION
**Pull request by clausmichele**
Here is my implementation of vector_to_random_points.
I've tried to include all the edge cases but there are still a couple of things to add, see the TODO in the code.

Sample code to test the implementation:

```
import openeo_processes as oeop

data = {
  'type': 'FeatureCollection',
  'features': [
    {
      'type': 'Feature',
      'properties': {},
      'geometry': {
        'type': 'Polygon',
        'coordinates': [
            [
              [
                11.12417221069336,
                46.09228143052647
              ],
              [
                11.171035766601562,
                46.09228143052647
              ],
              [
                11.171035766601562,
                46.11525271077457
              ],
              [
                11.12417221069336,
                46.11525271077457
              ],
              [
                11.12417221069336,
                46.09228143052647
              ]
            ]
        ]
      }
    },
    {
      'type': 'Feature',
      'properties': {},
      'geometry': {
        'type': 'MultiPolygon',
        'coordinates': [
          [
          [
            [
              11.14614486694336,
              46.06215432178359
            ],
            [
              11.16382598876953,
              46.06215432178359
            ],
            [
              11.16382598876953,
              46.069419674968515
            ],
            [
              11.14614486694336,
              46.069419674968515
            ],
            [
              11.14614486694336,
              46.06215432178359
            ]
          ]
          ],
          [
          [
            [
              11.1676025390625,
              46.06155875862966
            ],
            [
              11.186313629150389,
              46.06155875862966
            ],
            [
              11.186313629150389,
              46.068824190192196
            ],
            [
              11.1676025390625,
              46.068824190192196
            ],
            [
              11.1676025390625,
              46.06155875862966
            ]
          ]
          ]
        ]
      }
    },
    {
      'type': 'Feature',
      'properties': {},
      'geometry': {
        'type': 'Point',
        'coordinates': [
          11.124343872070312,
          46.059176441769665
        ]
      }
    },
    {
      'type': 'Feature',
      'properties': {},
      'geometry': {
        'type': 'Point',
        'coordinates': [
          11.132755279541016,
          46.072039731663025
        ]
      }
    }
  ]
}
points = oeop.vector_to_random_points(data,seed=None,geometry_count=10)

print(points.to_json())
```

### CODE REVIEW

**Comment by LukeWeidenwalker** [comment]
Thanks - will aim to provide a review by EOD tomorrow!

File: src/openeo_processes/arrays.py [hunk 1]
```diff
@@ -1668,3 +1676,179 @@ def exec_xar(data, asc=True, nodata=None, dimension=None):
...
+            count = 0
+            points_list = []
+            for i in range(pts_each_polygon[idx]):
+                # generate a random x and y
+                x = xmin + random.random() * xext
+                y = ymin + random.random() * yext
+                p = Point(x, y)
+                if polygon.geometry.contains(p):  # check if point is inside geometry
+                    points_list.append(p)
+                    count += 1
```
Comment (ValentinaHutter) [inline src/openeo_processes/arrays.py:1795]: If the Point is not in the geometry, you would instead just add the point in the center of the geometry, but shouldn't it always be possible to find a random Point inside a Polygon?

File: src/openeo_processes/arrays.py [hunk 2]
```diff
@@ -1668,3 +1676,179 @@ def exec_xar(data, asc=True, nodata=None, dimension=None):
...
+class VectorToRandomPoints:
+
+    @staticmethod
+    def exec_num(data, geometry_count = None, total_count = None, seed = None):
+
+        if seed is not None:
+            random.seed(seed)
+
+        # Each feature must have a properties field, even if there is no property
+        for feature in data['features']:
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/arrays.py:1698]: Is this really necessary? I just tried removing the properties field from the sample data you've provided and commented out this block and the result is unchanged!

Comment (LukeWeidenwalker) [inline src/openeo_processes/arrays.py:1698]: Is this really necessary? I just tried removing the properties field from the sample data you've provided and commented out this block and the result is unchanged!

Comment (LukeWeidenwalker) [inline src/openeo_processes/arrays.py:1698]: Is this really necessary? I just tried removing the properties field from the sample data you've provided and commented out this block and the result is unchanged!

File: src/openeo_processes/arrays.py [hunk 3]
```diff
@@ -1668,3 +1676,179 @@ def exec_xar(data, asc=True, nodata=None, dimension=None):
...
+                points_gdf = points_gdf.append(multipoint,ignore_index=True)
+
+            elif len(points_list)==1:
+                multipoint.geometry = points_list[0]
+                points_gdf = points_gdf.append(multipoint,ignore_index=True)
+
+
+        # Loop over all the MultiPolygons:
+        # polygons_from_multipolygons_gdf = multipolygons_gdf.explode()
+        for idx, polygon in multipolygons_gdf.iterrows():
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/arrays.py:1818]: In the current form the logic in this loop is exactly the same as the loop for Polygons above - @clausmichele are you aware of any edgecase handling that wants to be different for Polygons/Multipolygons? Otherwise I'd suggest folding this into one loop to save some lines of code!

Comment (LukeWeidenwalker) [inline src/openeo_processes/arrays.py:1818]: In the current form the logic in this loop is exactly the same as the loop for Polygons above - @clausmichele are you aware of any edgecase handling that wants to be different for Polygons/Multipolygons? Otherwise I'd suggest folding this into one loop to save some lines of code!

Comment (LukeWeidenwalker) [inline src/openeo_processes/arrays.py:1818]: In the current form the logic in this loop is exactly the same as the loop for Polygons above - @clausmichele are you aware of any edgecase handling that wants to be different for Polygons/Multipolygons? Otherwise I'd suggest folding this into one loop to save some lines of code!

**Review by LukeWeidenwalker — CHANGES_REQUESTED** [review]
Nice - I've left a few comments above, but the overall structure looks really good to me!

I know you're already aware of this, but thought I'd mention it here again: 
@soxofaan mentions the shapely representative_point function in the [other thread](https://github.com/Open-EO/openeo-processes/pull/315#issuecomment-1063031313) - I haven't had time to test it out yet, but this sounds quite promising indeed.

File: src/openeo_processes/arrays.py (hunk 1 above)
Comment (clausmichele) [reply src/openeo_processes/arrays.py:1795]: In theory yes, but how much time would it take? With complex geometries it's not obvious to easily find points within them. You can try to define a complex geometry by yourself using geojson.io and see.

File: src/openeo_processes/arrays.py (hunk 2 above)
Comment (clausmichele) [reply src/openeo_processes/arrays.py:1698]: Thanks for pointing this out. Actually I've also forgot to check if the property is None. I'll add it now.

It is related to this geopandas bug. I've opened a PR for that https://github.com/geopandas/geopandas/pull/2243

- If the geoJSON is valid (every geometry has a property field) and the properties are not `null` -> no problems
- If the geoJSON is valid (every geometry has a property field) and one property is `null` -> geopandas error
- If the geoJSON is missing a property field -> geopandas error

So the code is solving the second and third scenarios.

Comment (clausmichele) [reply src/openeo_processes/arrays.py:1698]: You can try the code again with my latest commit.

If you set one property to None in the Python dict and you comment that part, you'll get an error.

File: src/openeo_processes/arrays.py (hunk 3 above)
Comment (LukeWeidenwalker) [reply src/openeo_processes/arrays.py:1820]: Ah, yes, I understand that, I meant that specifically the point-generation loops for the Polygons (ln1780-1815) and then the MultiPolygons (ln1818-1856) seem to be doing the exact same thing, so couldn't we do something like this?

```suggestion
        for idx, polygon in pd.concat(polygons_gdf, multipolygons_gdf).iterrows():
```

More of a nitpick though!

**Comment by LukeWeidenwalker** [comment]
@clausmichele 
> I'm completely overwhelmed by this PR and can't really follow the status any longer.
Quoting Matthias from the other thread - I feel very much the same.

I'd propose we:
- merge this in as it is, because it does what it advertises. I'm happy with it already, we don't have to resolve my nitpick comment right now
- create a new issue to investigate using the shapely representative_point function instead in the future
- follow the new tickets around how this process needs to change in the future and write tests once the requirements are more stable 

Quick thumbs-up would be appreciated!

**Comment by clausmichele** [comment]
Let's merge it first, we'll try to follow the new proposals later on!
